k2/csrc/connect.cu

```diff
@@ -0,0 +1,497 @@
+        (int32_t arcs_idx012)->void {
+          // note: the prefix `arcs_` means it is an idxXXX w.r.t. `arcs_shape`.
+          // the prefix `fsas_` means the variable is an idxXXX w.r.t. `fsas_`.
+          int32_t arcs_idx01 = arcs_row_ids2_data[arcs_idx012],
+                  arcs_idx01x = arcs_row_splits2_data[arcs_idx01],
+                  arcs_idx2 = arcs_idx012 - arcs_idx01x,
+                  fsas_idx01 = states_data[arcs_idx01],  // a state index
+                  fsas_idx0 = fsas_row_ids1_data[fsas_idx01],
+                  fsas_idx01x = incoming_arcs_row_splits2_data[fsas_idx01],
+                  fsas_idx012 = fsas_idx01x + arcs_idx2,
+                  fsas_src_state_idx1 =
+                    fsas_data[incoming_arcs_data[fsas_idx012]].src_state,
+                  fsas_src_state_idx0x = fsas_row_splits1_data[fsas_idx0],
+                  fsas_src_state_idx01 =
+                    fsas_src_state_idx0x + fsas_src_state_idx1;
+          // if this arc is a self-loop, just ignore this arc as we won't
+          // processe the src_state (current state) again.
+          if (fsas_src_state_idx01 == fsas_idx01 ||
+              coaccessible_data[fsas_src_state_idx01]) {
+            keep_arc_data[arcs_idx012] = 0;
+            return;
+          }
+          keep_arc_data[arcs_idx012] = 1;
+          next_iter_states_data[arcs_idx012] = fsas_src_state_idx01;
+        });
+
+    Array1<int32_t> new2old_map = arc_renumbering.New2Old();
+    if (new2old_map.Dim() == 0) {
+      // There are no new states.  This means we terminated.  We'll check from
+      // calling code that we processed all arcs.
+      return nullptr;
+    }
+    // `new_states` will contain state-ids which are idx01's into `fsas_`.
+    Array1<int32_t> new_states = next_iter_states[new2old_map];
+    Array1<int32_t> new_states_row_ids(c_, new_states.Dim());  // will map to
+                                                               // FSA index
+    const int32_t *new2old_map_data = new2old_map.Data();
+    int32_t *new_states_row_ids_data = new_states_row_ids.Data();
+    K2_EVAL(
+        c_, new_states.Dim(), lambda_set_row_ids,
+        (int32_t new_state_idx)->void {
+          int32_t arcs_idx012 = new2old_map_data[new_state_idx],
+                  arcs_idx01 = arcs_row_ids2_data[arcs_idx012],  // state index
+              arcs_idx0 = arcs_row_ids1_data[arcs_idx01];        // FSA index
+          new_states_row_ids_data[new_state_idx] = arcs_idx0;
+        });
+
+    int32_t num_fsas = fsas_.Dim0();
+    Array1<int32_t> new_states_row_splits(c_, num_fsas + 1);
+    RowIdsToRowSplits(new_states_row_ids, &new_states_row_splits);
+
+    std::unique_ptr<Ragged<int32_t>> ans = std::make_unique<Ragged<int32_t>>(
+        RaggedShape2(&new_states_row_splits, &new_states_row_ids,
+        new_states.Dim()), new_states);
+    // The following will ensure the answer has deterministic numbering
```

Comment:
Is `deterministic` required by `Connect()`?
What if it is removed?

Reply:
Yeah, as I said about,  what we need in `GetNextBatchBackward` iteration is just `coaccessible_`, so I belive we can remove this line.